### PR TITLE
Vehicle and tollways data generator

### PR DESCRIPTION
**Pull request by BenGriffith**
Still have some more work to do here. So far:

1. Function creating fake vehicle information
2. Function/helpers scrapes Wikipedia webpage to generate a dictionary of US states and the tollways in each state. I had to get a little creative with approach due to webpage layout. 
3. Scraper is returning binary in a few spots. Planning to fix it.

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -0,0 +1,13 @@
+beautifulsoup4==4.12.2
+certifi==2023.7.22
+charset-normalizer==3.3.1
+Faker==19.12.0
```
Comment (bbelderbos) [inline requirements.txt:4]: no pip-tools + requirements.in or poetry + lock file?

File: tollway/vehicle.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+from us.states import STATES_AND_TERRITORIES
+
+
+STATE_NAMES = [state.name for state in STATES_AND_TERRITORIES]
+TOLLWAYS_URL = "https://en.wikipedia.org/wiki/List_of_toll_roads_in_the_United_States"
+
+
+def create_vehicle(fake: Faker) -> dict:
+    
+    vehicle = {f"{key.lower()}": value for key, value in fake.vehicle_object().items()}
```
Comment (bbelderbos) [inline tollway/vehicle.py:17]: seems f-string is not needed?

File: tollway/vehicle.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+    state_name = ""
+    for row in soup.contents[0]:
+        if not isinstance(row, Tag):
+            continue
+
+        if isinstance(row, Tag) and row.name == "h2" and row.span.string in STATE_NAMES:
+            state_name = row.span.string
+            table_count_between_states = 0
+
+        if isinstance(row, Tag) and row.name == "table" and row["class"][0] == "wikitable" and state_name != "":
```
Comment (bbelderbos) [inline tollway/vehicle.py:38]: you can simplify this with a helper function, also some duplication with the if above

Comment (bbelderbos) [reply tollway/vehicle.py:38]: is the isinstance check needed? you already continued when it's not an instance of Tag

File: tollway/vehicle.py [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+
+def _state_tollways(soup: BeautifulSoup) -> list:
+    names = []
+
+    for table in soup.find_all("table", class_="wikitable"):
+        table_rows = table.tbody.find_all("tr")[1:]
+        state_tollways = []
+        for table_row in table_rows:
+            for tag in table_row.find_all("a"):
```
Comment (bbelderbos) [inline tollway/vehicle.py:52]: listcomp maybe?

```
state_tollways = [tag.string for tag in table_row.find_all("a") if tag.string is not None]
```

File: tollway/vehicle.py [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+                    continue
+                state_tollways.append(tag.string)
+                break
+        names.append(state_tollways)
+    
+    return names
+
+
+def get_tollways() -> dict:
+    tollways_html = requests.get(TOLLWAYS_URL)
```
Comment (bbelderbos) [inline tollway/vehicle.py:63]: you could pass the html into the function as arg to decouple the request call, makes it easier to test

File: tollway/vehicle.py [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+    tollway_content = wikipedia_soup.find("div", class_="mw-parser-output")
+    tollway_subsection = BeautifulSoup(markup=str(tollway_content), features="html.parser")
+
+    num_of_tables_between_states = _num_of_tables_between_states(soup=tollway_subsection)
+    state_tollways = deque(_state_tollways(soup=tollway_subsection))
+
+    united_states_tollways = {}
+    for state_name, table_count in num_of_tables_between_states.items():
+
+        merge_tollway_tables = []
```
Comment (bbelderbos) [inline tollway/vehicle.py:75]: why do you need these 3 lines?

File: tollway/vehicle.py [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+
+        united_states_tollways[state_name] = set(merge_tollway_tables)
+        merge_tollway_tables = []
+
+    return united_states_tollways
+
+
+def create_tollway(tollways: dict) -> [str, str]:
+    state = choice([key for key in tollways.keys()])
+    name = choice(list(tollways[state]))
```
Comment (bbelderbos) [inline tollway/vehicle.py:87]: might not need the list() cast, I think choice is happy with an iterable

File: tollway/vehicle.py [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+            merge_tollway_tables.extend(state_tollways.popleft())
+
+        united_states_tollways[state_name] = set(merge_tollway_tables)
+        merge_tollway_tables = []
+
+    return united_states_tollways
+
+
+def create_tollway(tollways: dict) -> [str, str]:
+    state = choice([key for key in tollways.keys()])
```
Comment (bbelderbos) [inline tollway/vehicle.py:86]: can you just give choice() tollways.keys(), seems the listcomp does not add any value

File: tollway/vehicle.py [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+    
+    return names
+
+
+def get_tollways() -> dict:
+    tollways_html = requests.get(TOLLWAYS_URL)
+    wikipedia_soup = BeautifulSoup(markup=tollways_html.text, features="html.parser")
+
+    tollway_content = wikipedia_soup.find("div", class_="mw-parser-output")
+    tollway_subsection = BeautifulSoup(markup=str(tollway_content), features="html.parser")
```
Comment (bbelderbos) [inline tollway/vehicle.py:67]: tollway_content is already a soup like object and should have all relevant methods, do we need to make a new BeautifulSoup object here? I usually only have one in my bs4 scripts so the one you have on L64

**Review by bbelderbos — COMMENTED** [review]
Good start, please check my feedback ...

File: requirements.txt (hunk 1 above)
Comment (BenGriffith) [reply requirements.txt:4]: Thanks for pointing this out. I'll add pip-tools.

File: tollway/vehicle.py (hunk 2 above)
Comment (BenGriffith) [reply tollway/vehicle.py:17]: yes, it isn't needed. Good catch.

File: tollway/vehicle.py (hunk 8 above)
Comment (BenGriffith) [reply tollway/vehicle.py:86]: I replaced with list() cast, when I removed the listcomp was receiving type error saying dict_keys object is not subscriptable

File: tollway/vehicle.py (hunk 9 above)
Comment (BenGriffith) [reply tollway/vehicle.py:67]: good to know you usually only have one. I've gone back and forth on this.

File: tollway/vehicle.py (hunk 8 above)
Comment (bbelderbos) [reply tollway/vehicle.py:86]: do you need the cast? keys() is an iterable so that probably just works out of the box?

Comment (bbelderbos) [reply tollway/vehicle.py:86]: also `[str, str]` type hint issue, should this be `dict[str, str]`?

Comment (BenGriffith) [reply tollway/vehicle.py:86]: it is returning two strings which is what I was trying to communicate with `[str, str]` however that syntax might be incorrect and instead communicate a list with two strings is expected

Comment (BenGriffith) [reply tollway/vehicle.py:86]: this is what the output/return looks like in it's current state

Screenshot 2023-11-07 at 7 32 06 AM

Comment (BenGriffith) [reply tollway/vehicle.py:86]: this is what the return/output looks like in it's current state

Screenshot 2023-11-07 at 7 33 07 AM

Comment (BenGriffith) [reply tollway/vehicle.py:86]: so maybe I need to use parens to denote tuple instead of brackets

Comment (bbelderbos) [reply tollway/vehicle.py:86]: thanks for confirming that, then it would be `tuple[str, str]:`

**Pull request merged**